File: clients/prover/vault.py
"""The on-device vault file.

What this file holds is the whole reason the server cannot do the client's
job: four plaintext figures and the key that authorises saying anything about
them. Everything else in Memtara is designed so that this file never leaves
the device it was created on.

-------------------------------------------------------------------
THIS IS NOT WHAT A SHIPPING PRODUCT USES
-------------------------------------------------------------------
A JSON file on disk is the right shape for an advisor terminal, a CI run and
a demonstration, and the wrong shape for a phone. In the mobile app the
signing key belongs in the Secure Enclave / Android Keystore, where it can be
used but not exported, and the figures belong in the encrypted vault that
`vault/` (the Rust crate) already implements. The format here deliberately
mirrors that structure — a key, and a category of figures — so the port is a
substitution rather than a redesign.

Because it is plaintext, this module refuses to load a vault whose file mode
is readable by anyone but its owner. That check is cheap, it catches the
single most likely real mistake (a vault committed to a repository or dropped
in /tmp), and a warning nobody reads would not.
"""

from __future__ import annotations

import json
import os
import secrets
import stat
import sys
from dataclasses import dataclass
from pathlib import Path
from typing import Any

sys.path.insert(0, str(Path(__file__).resolve().parents[1]))

from wealth_client import SUBORDER, Keypair, WealthVault  # noqa: E402

VAULT_VERSION = 1

#: The four figures a suitability assessment reads. Named here so a typo in a
#: hand-edited vault fails with "unknown figure 'liquid_asets'" instead of
#: silently proving against a zero.
FIGURE_NAMES = ("income", "liquid_assets", "risk_tolerance", "existing_holdings_value")


class VaultError(RuntimeError):
    pass


@dataclass(frozen=True)
class LoadedVault:
    path: Path
    user_id: str | None
    wealth: WealthVault

    @property
    def keypair(self) -> Keypair:
        return self.wealth.keypair


def create(
    path: Path,
    *,
    income: int,
    liquid_assets: int,
    risk_tolerance: int,
    existing_holdings_value: int,
    user_id: str | None = None,
    seed: int | None = None,
    overwrite: bool = False,
) -> LoadedVault:
    """Write a new vault.

    ``seed`` exists so a demo or a test can reproduce a client exactly. It
    defaults to fresh CSPRNG output, and a caller supplying one is asserting
    that this vault is not protecting anything real — the CLI says as much
    when the flag is used.
    """
    if path.exists() and not overwrite:
        raise VaultError(f"{path} already exists; pass --overwrite to replace it")

    _validate_figures(income, liquid_assets, risk_tolerance, existing_holdings_value)

    if seed is None:
        seed = secrets.randbelow(SUBORDER - 1) + 1
    keypair = Keypair.from_seed(seed)

    document: dict[str, Any] = {
        "version": VAULT_VERSION,
        "user_id": user_id,
        # Stored as the seed, not as the derived scalar. `Keypair.from_seed`
        # applies the cofactor-8 multiplication that `eddsa_verify` expects,
        # and storing the post-cofactor scalar would invite someone to
        # reconstruct the keypair without it — producing signatures that fail
        # inside the circuit with no diagnostic beyond "unsatisfied
        # constraint".
        "signing_key_seed": format(seed, "064x"),
        "figures": {
            "income": income,
            "liquid_assets": liquid_assets,
            "risk_tolerance": risk_tolerance,
            "existing_holdings_value": existing_holdings_value,
        },
    }

    path.parent.mkdir(parents=True, exist_ok=True)
    # Created 0600 from the start rather than written and then chmod'ed:
    # between those two calls the file is world-readable, and on a shared
    # advisor terminal that window is enough.
    fd = os.open(path, os.O_WRONLY | os.O_CREAT | os.O_TRUNC, 0o600)
    with os.fdopen(fd, "w", encoding="utf-8") as handle:
        json.dump(document, handle, indent=2, sort_keys=True)
        handle.write("\n")

    return load(path)


def load(path: Path, *, allow_insecure_permissions: bool = False) -> LoadedVault:
    if not path.is_file():
        raise VaultError(f"no vault at {path}")

    if not allow_insecure_permissions and os.name == "posix":
        mode = stat.S_IMODE(path.stat().st_mode)
        if mode & 0o077:
            raise VaultError(
                f"{path} is mode {mode:04o} — it contains plaintext income and portfolio figures "
                f"and a signing key, and is readable beyond its owner. Run "
                f"`chmod 600 {path}`, or pass --allow-insecure-permissions if this vault is "
                f"disposable test data"
            )

    try:
        document = json.loads(path.read_text(encoding="utf-8"))
    except json.JSONDecodeError as exc:
        raise VaultError(f"{path} is not valid JSON: {exc}") from None

    version = document.get("version")
    if version != VAULT_VERSION:
        raise VaultError(f"{path} declares vault version {version!r}; this tool understands {VAULT_VERSION}")

    figures = document.get("figures")
    if not isinstance(figures, dict):
        raise VaultError(f"{path} has no `figures` object")
    unknown = sorted(set(figures) - set(FIGURE_NAMES))
    if unknown:
        # Refused rather than ignored. A vault carrying `liquid_asets` would
        # otherwise prove against a liquid_assets of zero and fail the
        # assessment for a reason nobody could see.
        raise VaultError(f"{path} has unknown figures: {', '.join(unknown)}")
    missing = sorted(set(FIGURE_NAMES) - set(figures))
    if missing:
        raise VaultError(f"{path} is missing figures: {', '.join(missing)}")

    values = {name: figures[name] for name in FIGURE_NAMES}
    for name, value in values.items():
        if not isinstance(value, int) or isinstance(value, bool):
            raise VaultError(f"{path}: `{name}` must be an integer, got {value!r}")
    _validate_figures(**values)

    seed_hex = document.get("signing_key_seed")
    if not isinstance(seed_hex, str):
        raise VaultError(f"{path} has no `signing_key_seed`")
    try:
        seed = int(seed_hex, 16)
    except ValueError:
        raise VaultError(f"{path}: `signing_key_seed` is not hexadecimal") from None

    return LoadedVault(
        path=path,
        user_id=document.get("user_id"),
        wealth=WealthVault(keypair=Keypair.from_seed(seed), **values),
    )


def _validate_figures(income: int, liquid_assets: int, risk_tolerance: int, existing_holdings_value: int) -> None:
    for name, value in (
        ("income", income),
        ("liquid_assets", liquid_assets),
        ("existing_holdings_value", existing_holdings_value),
    ):
        if value < 0:
            raise VaultError(f"`{name}` must not be negative")
        # The circuit takes u64. A figure above that is not a vault a proof
        # can be generated from, and finding out at `nargo execute` time costs
        # the user a confusing failure several seconds later.
        if value >= 2**64:
            raise VaultError(f"`{name}` exceeds the u64 the circuit accepts")
    if not 1 <= risk_tolerance <= 5:
        # Asserted inside the circuit too. Caught here so a mistyped
        # questionnaire answer is a readable error rather than an
        # unsatisfiable constraint.
        raise VaultError("`risk_tolerance` must be between 1 and 5 (the scale the circuit asserts)")

File: clients/web-prover/src/vault.js
// The in-browser reading of a vault file — the JS-side counterpart of
// `clients/prover/vault.py`. Same format, same four figures, same signing
// key, same refusal to guess at a malformed field: a vault carrying
// `liquid_asets` fails loudly here exactly as it does there, rather than
// silently proving against a liquid_assets of zero.
//
// The one thing this module does that the Python one cannot: it never touches
// a filesystem. The caller hands it the `File` object straight from an
// `<input type="file">`, it is read with `FileReader` entirely in page
// memory, and the parsed figures and signing key never leave this module
// except as arguments to the local proving pipeline in `prove.js`. See
// `app.js` for the file input wiring and the UI copy that says as much.

export const VAULT_VERSION = 1;
export const FIGURE_NAMES = ['income', 'liquid_assets', 'risk_tolerance', 'existing_holdings_value'];

// Leaf slots in the wealth category of the vault tree — fixed positions,
// matching circuits/wealth_suitability/src/main.nr's test module and
// clients/wealth_client.py's INCOME_SLOT / LIQUID_SLOT / RISK_SLOT /
// HOLDINGS_SLOT.
export const INCOME_SLOT = 0;
export const LIQUID_SLOT = 1;
export const RISK_SLOT = 2;
export const HOLDINGS_SLOT = 3;
export const TREE_LEAVES = 16;

export class VaultError extends Error {}

function isPlainInteger(v) {
  return typeof v === 'number' && Number.isInteger(v);
}

function validateFigures({ income, liquid_assets, existing_holdings_value, risk_tolerance }) {
  for (const [name, value] of [
    ['income', income],
    ['liquid_assets', liquid_assets],
    ['existing_holdings_value', existing_holdings_value],
  ]) {
    if (value < 0) throw new VaultError(`\`${name}\` must not be negative`);
    // The circuit takes u64. A figure above that is not a vault a proof can
    // be generated from — caught here, exactly as vault.py's
    // `_validate_figures` catches it, rather than several seconds into
    // circuit execution.
    if (BigInt(value) >= 2n ** 64n) throw new VaultError(`\`${name}\` exceeds the u64 the circuit accepts`);
  }
  if (!(risk_tolerance >= 1 && risk_tolerance <= 5)) {
    throw new VaultError('`risk_tolerance` must be between 1 and 5 (the scale the circuit asserts)');
  }
}

/** Parse and validate a vault document already decoded from JSON (the
 * `{version, user_id, signing_key_seed, figures}` shape `clients/prover/
 * vault.py` writes). Returns `{userId, figures, signingKeySeed}` with
 * `signingKeySeed` as a BigInt, ready for `babyjubjub.keypairFromSeed`. */
export function parseVaultDocument(document) {
  if (document.version !== VAULT_VERSION) {
    throw new VaultError(`vault declares version ${JSON.stringify(document.version)}; this page understands ${VAULT_VERSION}`);
  }
  const figures = document.figures;
  if (typeof figures !== 'object' || figures === null || Array.isArray(figures)) {
    throw new VaultError('vault has no `figures` object');
  }
  const present = Object.keys(figures);
  const unknown = present.filter((k) => !FIGURE_NAMES.includes(k));
  if (unknown.length > 0) {
    throw new VaultError(`vault has unknown figures: ${unknown.sort().join(', ')}`);
  }
  const missing = FIGURE_NAMES.filter((k) => !present.includes(k));
  if (missing.length > 0) {
    throw new VaultError(`vault is missing figures: ${missing.sort().join(', ')}`);
  }
  const values = {};
  for (const name of FIGURE_NAMES) {
    const v = figures[name];
    if (!isPlainInteger(v)) {
      throw new VaultError(`vault: \`${name}\` must be an integer, got ${JSON.stringify(v)}`);
    }
    values[name] = v;
  }
  validateFigures(values);

  const seedHex = document.signing_key_seed;
  if (typeof seedHex !== 'string' || seedHex.length === 0) {
    throw new VaultError('vault has no `signing_key_seed`');
  }
  let signingKeySeed;
  try {
    signingKeySeed = BigInt('0x' + seedHex.replace(/^0x/i, ''));
  } catch {
    throw new VaultError('vault: `signing_key_seed` is not hexadecimal');
  }

  return {
    userId: document.user_id ?? null,
    figures: values,
    signingKeySeed,
  };
}

/** Read a `File` (from a file input or drag-and-drop) with `FileReader` and
 * parse it as a vault document. Never touches `fetch`/`XMLHttpRequest` — the
 * bytes go from disk into page memory and no further, which is the whole
 * point of this function existing separately from a generic JSON fetch. */
export function readVaultFile(file) {
  return new Promise((resolve, reject) => {
    const reader = new FileReader();
    reader.onerror = () => reject(new VaultError(`could not read ${file.name}: ${reader.error}`));
    reader.onload = () => {
      let document;
      try {
        document = JSON.parse(reader.result);
      } catch (e) {
        reject(new VaultError(`${file.name} is not valid JSON: ${e.message}`));
        return;
      }
      try {
        resolve(parseVaultDocument(document));
      } catch (e) {
        reject(e);
      }
    };
    reader.readAsText(file);
  });
}

/** The 16-leaf array the vault tree commits to: the four wealth figures in
 * their reserved slots, zero everywhere else — mirrors
 * `WealthVault.leaves()`. */
export function vaultLeaves(figures) {
  const leaves = new Array(TREE_LEAVES).fill(0n);
  leaves[INCOME_SLOT] = BigInt(figures.income);
  leaves[LIQUID_SLOT] = BigInt(figures.liquid_assets);
  leaves[RISK_SLOT] = BigInt(figures.risk_tolerance);
  leaves[HOLDINGS_SLOT] = BigInt(figures.existing_holdings_value);
  return leaves;
}

/** What the circuit should conclude, evaluated independently of it —
 * mirrors `WealthVault.expected_verdict`. Used the same way
 * `clients/prover/cli.py` uses the Python original: as a second opinion the
 * proving pipeline refuses to disagree with silently. */
export function expectedVerdict(figures, terms) {
  const total = figures.liquid_assets + figures.existing_holdings_value;
  return (
    figures.income >= terms.minIncome &&
    figures.liquid_assets >= terms.minLiquidity &&
    figures.risk_tolerance >= terms.productRiskLevel &&
    figures.existing_holdings_value * 100 <= terms.maxConcentrationPercent * total
  );
}
